### ISSUE TITLE
When using certificates, the generated ID is not always correctly matching the SAN

### ISSUE THREAD

**Issue by letoams** [linux-system-roles/vpn#19]
I used certificates and got a generated connection file:

conn 192.1.2.23-to-192.1.2.45
  left=192.1.2.23
  leftid=@192.1.2.23
  right=192.1.2.45
  ikev2=insist
  auto=ondemand
  rightid=%fromcert
  leftrsasigkey=%cert
  leftcert=east
  rightrsasigkey=%cert

Using leftid=@192.1.2.23 means that a SAN of type DNS with the name "192.1.2.23" has to exist on the cert, which will not be case. If there is a SAN, it will be of type IP.

If leftid= is an IP and the same as the left= it should either be specified as leftid=192.1.2.23 or omitted and libreswan will create an ID based on the IP itself.

The result is:

"192.1.2.45-to-192.1.2.23" #1: certificate verified OK: E=user-west@testing.libreswan.org,CN=west.testing.libreswan.org,OU=Test Department,O=Libreswan,L=Toronto,ST=Ontario,C=CA
"192.1.2.45-to-192.1.2.23" #1: IKEv2 mode peer ID is ID_FQDN: '@192.1.2.45'
"192.1.2.45-to-192.1.2.23" #1: authenticated using RSA with SHA2_512
"192.1.2.45-to-192.1.2.23" #1: responding to IKE_AUTH message (ID 1) from 192.1.2.45:500 with encrypted notification AUTHENTICATION_FAILED

Note authentication of the certificate succeeded but the ID_FQDN used as IKE peer ID was not a SAN on that certificate, so it rejected authentication.

my inventory file looked like:

 all:
  hosts:
    west:
      ansible_host: 192.1.2.45
      ansible_user: root
      ansible_password: swan
      ansible_ssh_common_args: -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no
    east:
      ansible_host: 192.1.2.23
      ansible_user: root
      ansible_password: swan
      ansible_ssh_common_args: -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no
  vars:
    vpn_auth_method: cert
    vpn_connections:
      - hosts:
          west:
            cert_name: west
          east:
            cert_name: east
        auto: ondemand

**Comment by mprovenc** [comment]
@letoams Are there any related changes that need to be applied to the .secrets file? Should the `@` be omitted from the left value in the .secrets file if `left` is an IP address? Also, should this scheme be made generic for any auth method or only apply when certs are used?

**Comment by letoams** [comment]
On Mon, 29 Mar 2021, mprovenc wrote:

> @letoams Are there any related changes that need to be applied to the .secrets file? Should the @ be omitted from the left value in the
> .secrets file if left is an IP address? Also, should this scheme be made generic for any auth method or only apply when certs are used?

You should never generate an @IP_ADDRESS, because the @ symbol means
IDtype of ID_FQDN. Omitting the @ means IDtype is ID_IPv4/ID_IPv6.

Ideally, certificates have a proper SAN with a FQDN in it, so you
would use that as ID_FQDN. So if you have a certificate for
west.libreswan.org, you might have a CN=west.libreswan.org (but CN
is ignored for verification) and a SAN of DNS:west.libreswan.org,
so you specify ***@***.***

If the CN has something like "Paul Wouters", then you either use the
full DN of the cert, of a full DN with wildcards. So it depends a bit
on how the certificates are generated. I dont expect this module to
do any parsing/checking of the certificate though.

So in short, if the ID value is an IPv4 or IPv6 address, strip the "@",
for both PSK and CERT authby methods
